### ISSUE TITLE
Renaming folder actually deletes most of its contents

### ISSUE THREAD

**Issue by RobinJ1995** [owncloud/client#3129]
I was on a holiday for about a week. I had no internet connection on my laptop, but added some pictures into my `Foto's` folder. I just connected to the internet, and renamed the folder in question to `Pictures`. The data I added to the folder is still there, everything that was in there before is gone, both on the server and on the client. That's ~10GB of pictures that just went poof.

Please look into this. Much appreciated.

![screenshot_2015-04-18_22-21-39](https://cloud.githubusercontent.com/assets/3089559/7217123/de07e5a8-e619-11e4-95e1-d17ed8d1f9c4.png)
![screenshot_2015-04-18_22-25-43](https://cloud.githubusercontent.com/assets/3089559/7217125/ee29109c-e619-11e4-9cf6-a5cc1c5518e3.png)
![screenshot_2015-04-18_22-28-00](https://cloud.githubusercontent.com/assets/3089559/7217132/313b5426-e61a-11e4-8c50-e09a62f25886.png)
![screenshot_2015-04-18_22-28-32](https://cloud.githubusercontent.com/assets/3089559/7217136/43db2516-e61a-11e4-9e04-d4c16b841237.png)

**Comment by guruz** [comment]
Ouch, this sounds bad.

@PVince81 @icewind1991 @DeepDiver1975 Can you help with this? From the screenshots it looks like server 8.0.2

@RobinJ1995 btw we also have https://raw.githubusercontent.com/owncloud/client/master/issue_template.md

**Comment by PVince81** [comment]
Big folder rename, likely to be: https://github.com/owncloud/core/issues/13391

If it is a shared folder that you renamed as recipient, and renamed it back to the original sharer name, then it's: https://github.com/owncloud/core/issues/15340

**Comment by RobinJ1995** [comment]
The folder in question was my own folder, and was not shared with anyone.

Regarding owncloud/core#13391;
I renamed the folder on my laptop using my file manager, not though the web client. I also didn't notice anything strange in the web client like a page that took a long time to load. The data was just gone.

I have access to the server in question, so if you tell me which log files (if any) you would like to see, I can provide them.

**Comment by PVince81** [comment]
This could also happen with your own folder if it contains many files and subfolders.
In the case of shared folders it would be even slower and increase the probability of the problem happening.

**Comment by RobinJ1995** [comment]
So is there any clue as to why this is happening? Is it perhaps checking stuff in the wrong order, or doing them at the same time?

**Comment by PVince81** [comment]
Yes, this happens because the index needs to be updated for a lot of entries. There is a window of time where the files on the server disk and index do not match. If during this time another client tries to access this folder, it will access outdated data and be confused. This can also happen if the time needed to update the database takes longer than a sync cycle, which means that while the PHP process is still running the next sync run will kick in. 

There is a workaround here https://github.com/owncloud/core/issues/13391#issuecomment-73243546 to make the DB update faster, but it won't eliminate the probability of concurrency.

In the future there will be high-level file locking https://github.com/owncloud/core/issues/11804 to make sure that such concurrent operations cannot cause trouble.

**Comment by PVince81** [comment]
with "confused" I mean: the second process will find that what's on disk doesn't match the cache, then it will ALSO try to update the index/cache with what's on disk. Basically there will be two processes contradicting themselves about what needs to be written in the index/cache.

**Comment by RobinJ1995** [comment]
Ah. So some kind of locking mechanism should be implemented :)
Thanks for the explanation.

My desktop computer still had the files, so I unplugged the network cable before booting, copied my ownCloud folder, and reconnected to the network. Then I proceeded to upload all my files again (which took a while). So at least (almost) no data was permanently lost for me, but others may not be as lucky.
I hope this issue gets resolved quickly.
